crates/resolver/src/error.rs

```diff
@@ -185,12 +185,12 @@ impl ResolveError {
             response_code @ ResponseCode::NoError
             if !response.contains_answer() && !response.truncated() => {
                 // TODO: if authoritative, this is cacheable, store a TTL (currently that requires time, need a "now" here)
-                // let valid_until = if response.is_authoritative() { now + response.get_negative_ttl() };
+                // let valid_until = if response.authoritative() { now + response.negative_ttl() };
 
                 let mut response = response;
                 let soa = response.soa().cloned();
                 let negative_ttl = response.negative_ttl();
-                let trusted = if response_code == ResponseCode::NoError { false } else { trust_nx };
+                let trusted = trust_nx && response.authoritative();
```

Comment:
Don't we want `||` here rather than `&&`? Seems like we should always trust an authoritative server, and only use the `trust_nx` bit for unauthoritative servers.

Reply:
Some untrustworthy servers might outright lie about being authoritative, and if they redact records we should try other servers. That's where my brain was going, but I think I actually misinterpreted the spec and it doesn't matter whether or not the return is authoritative. I'll revise this.

Reply:
I think this is more of a measure against incompetent servers rather than malicious ones, IIRC.

Reply:
Thinking about this a bit more, to fit in with how retries work we only want to say the lack of results is final IF there's a SOA record present. So for reverse lookups against private space e.g. `1.1.168.192.in-addr.arpa. PTR` you'll get NXDOMAIN from reasonable resolvers with no SOA, because if there are other resolvers (like local ones) they may have valid addresses elsewhere.

Improperly configured servers may try to do recursive lookups upstream for these, and they'd return the [AS112 blackhole SOAs](https://www.as112.net/). We should probably handle that case eventually, but I think it's out of scope for now.